### PR TITLE
Add resource path generators doc

### PR DESCRIPTION
**Pull request by polc**
| Q             | A
| ------------- | ---
| License       | MIT

I'm not used to write doc, comments welcome.

### CODE REVIEW

File: core/resource-path-generators.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
+# Resource path generators
+
+API Platform allow you to configure the path of your exposed resources.
+It come with pre registered generators but you can easily define your own.
```
Comment (soyuka) [inline core/resource-path-generators.md]: With API Platform, you can configure the path of your exposed resources. It comes with pre-registered generators, which you can override easily.

File: core/resource-path-generators.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
+# Resource path generators
+
+API Platform allow you to configure the path of your exposed resources.
+It come with pre registered generators but you can easily define your own.
+
+## Configuration
+
+There is 2 pre registered generator services :
```
Comment (soyuka) [inline core/resource-path-generators.md]: There are two pre-registered path generator services:

File: core/resource-path-generators.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+```yaml
+api_platform:
+    routing:
+        resource_path_generator: api_platform.routing.resource_path_generator.dash
+```
+
+
+## Create a custom path generator
+
+Let's assumes we need urls without separators, ex: `api.tld/myresources`
```
Comment (soyuka) [inline core/resource-path-generators.md]: Let's assume we need urls without separators (eg `api.tld/myresources`)

File: core/resource-path-generators.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+<services>
+    <service id="app.routing.resource_path_generator.no_separators" class="AppBundle\Routing\NoSeparatorsResourcePathGenerator" /> 
+</services>
+```
+
+</configurations>
+
+### Configure it
+
+```yaml
```
Comment (soyuka) [inline core/resource-path-generators.md]: Add `app/config/config.yml` as a comment so that we know where to add this to ;)

**Comment by polc** [comment]
Ready for review.

File: core/resource-path-generators.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+```
+
+
+## Create a custom path generator
+
+Let's assumes we need urls without separators (eg `api.tld/myresources`)
+
+### Defining the Resource Path Generator
+
+Make sure the custom generator extends `ApiPlatform\Core\Routing\ResourcePathGeneratorInterface` :
```
Comment (soyuka) [inline core/resource-path-generators.md]: implements not extends

File: core/resource-path-generators.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+With API Platform, you can configure the path of your exposed resources. 
+It comes with pre-registered generators, which you can override easily.
+
+## Configuration
+
+There are two pre-registered path generator services:
+
+| strategy                                                  | entity name  | path result                       |
+|-----------------------------------------------------------|--------------|-----------------------------------|
+| `api_platform.routing.resource_path_generator.underscore` | `MyResource` | `api.tld/my_resources`            |
```
Comment (dunglas) [inline core/resource-path-generators.md]: I would remove the `api.tld` part. If you really want to keep the domain name, please use `example.com` (it's a standard domain for testing and doc purpose).

**Comment by dunglas** [comment]
Looks good to me, I just left a small comment about the domain name.

File: core/resource-path-generators.md (hunk 6 above)
Comment (soyuka) [inline core/resource-path-generators.md]: As the symfony `debug:router` doesn't give any hostname I think it'd be better without.

**Comment by polc** [comment]
Updated

**Comment by soyuka** [comment]
👍

**Comment by Simperfit** [comment]
👍

File: core/index.md [hunk 7]
```diff
@@ -95,6 +95,9 @@ This bundle is documented and tested with Behat (take a look at [the `features/`
   1. [Enabling the metadata cache](performances.md#enabling-the-metadata-cache)
 14. [Security](security.md) 
 15. [AngularJS integration](angular-integration.md)
+16. [Resources path generators](resource-path-generator.md)
```
Comment (dunglas) [inline core/index.md]: Can you put it before the angular integration please (I just moved all integrations with external tools at the end of the TOC).

File: core/resource-path-generators.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    <service id="app.routing.resource_path_generator.no_separators" class="AppBundle\Routing\NoSeparatorsResourcePathGenerator" /> 
+</services>
+```
+
+</configurations>
+
+### Configure it
+
+```yaml
+# app/config/config.yml
```
Comment (dunglas) [inline core/resource-path-generators.md]: Can you add a blank line after the comment?

File: core/index.md (hunk 7 above)
Comment (dunglas) [inline core/index.md]: Can you also add subsections of your article in the TOC (like for previous entries)?

**Comment by dunglas** [comment]
Sorry I just made some more comments. It's almost ready to be merged!

**Comment by polc** [comment]
Is it ok ? I'm not sure about the Previous Chapter link, should I rebase ?

File: core/angular-integration.md [hunk 9]
```diff
@@ -96,3 +96,4 @@ nga.configure(admin)
 You can look at what we have done as another exemple [api-platform/admin](https://github.com/api-platform/admin).
 
 Previous chapter: [Performances](performances.md)
+Next chapter: [Resource path generators](resource-path-generators.md)
```
Comment (Simperfit) [inline core/angular-integration.md]: I think this can be after point 6 it will make more sens then after angular :p

File: core/index.md [hunk 10]
```diff
@@ -94,6 +94,9 @@ This bundle is documented and tested with Behat (take a look at [the `features/`
 13. [Performances](performances.md)
   1. [Enabling the metadata cache](performances.md#enabling-the-metadata-cache)
 14. [Security](security.md) 
+15. [Resources path generators](resource-path-generator.md)
```
Comment (Simperfit) [inline core/index.md]: 16

File: core/index.md [hunk 11]
```diff
@@ -94,6 +94,9 @@ This bundle is documented and tested with Behat (take a look at [the `features/`
...
+  1. [Configuration](resource-path-generator.md#configuration)
+  2. [Create a custom path generator](resource-path-generator.md#create-a-custom-path-generator)
 15. [AngularJS integration](angular-integration.md)
```
Comment (Simperfit) [inline core/index.md]: 16

File: core/resource-path-generators.md [hunk 12]
```diff
@@ -0,0 +1,89 @@
+# Resource path generators
+
+With API Platform, you can configure the path of your exposed resources. 
+It comes with pre-registered generators, which you can override easily.
```
Comment (Simperfit) [inline core/resource-path-generators.md]: `you can easily override `

File: core/resource-path-generators.md [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+```yaml
+api_platform:
+    routing:
+        resource_path_generator: api_platform.routing.resource_path_generator.dash
+```
+
+
+## Create a custom path generator
+
+Let's assumes we need urls without separators (eg `api.tld/myresources`)
```
Comment (theofidry) [inline core/resource-path-generators.md]: `e.g.`

File: core/resource-path-generators.md [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+// src/AppBundle/Routing/NoSeparatorsResourcePathGenerator.php
+
+namespace AppBundle\Routing;
+
+use ApiPlatform\Core\Routing\ResourcePathGeneratorInterface;
+use Doctrine\Common\Inflector\Inflector;
+
+class NoSeparatorsResourcePathGenerator implements ResourcePathGeneratorInterface
+{
+    public function generateResourceBasePath(string $resourceShortName) : string
```
Comment (theofidry) [inline core/resource-path-generators.md]: What's Symfony CS regarding semi-columns for the typehint @dunglas? Personally I'm not putting one as in Hack...

File: core/resource-path-generators.md [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+    {
+        $pathName = strtolower($resourceShortName);
+
+        return Inflector::pluralize($pathName);
+    }
+}
+```
+
+Note that `$resourceShortName` contains a camel case string, by default the resource class name (eg `MyResource`).
```
Comment (theofidry) [inline core/resource-path-generators.md]: `e.g`

File: core/resource-path-generators.md [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+
+### Registering the service
+
+<configurations>
```
Comment (theofidry) [inline core/resource-path-generators.md]: what is that `configuration` tag for?

Comment (soyuka) [inline core/resource-path-generators.md]: I think it's for the tabs to change between yaml/xml/annotations

**Comment by Simperfit** [comment]
Could you please rebase with the master branch ?

Thanks for this work !

File: core/resource-path-generators.md (hunk 14 above)
Comment (soyuka) [inline core/resource-path-generators.md]: You mean colon? It's correct having space before and after.

Comment (theofidry) [inline core/resource-path-generators.md]: Yeah, just asking which one are enforcing

Comment (theofidry) [inline core/resource-path-generators.md]: nvm just had to look at api-platform/core for the answer

**Comment by polc** [comment]
Re-based !

**Comment by theofidry** [comment]
👌

**Comment by dunglas** [comment]
Thank you @polc!

**Pull request merged**